x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/ldap/PoolingSessionFactorySettings.java

```diff
@@ -35,6 +36,26 @@
                     key -> secureString(key, null)
             );
 
+    public static final Function<String, Setting.AffixSetting<String>> BIND_MODE = RealmSettings.affixSetting("bind.mode",
+            key -> new Setting<>(key, "simple", Function.identity(), v -> {
+                switch (v) {
+                case "simple":
+                case "sasl_gssapi":
+                    break;
+                default:
+                    throw new IllegalArgumentException("only [simple] and [sasl_gssapi] bind mode are allowed, [" + v + "] is invalid");
+                }
+            }, Setting.Property.NodeScope));
```

Comment:
I like the _switch_ validation !
But, can't we infer the mode from if `SASL_GSSAPI_PRINCIPAL` is present. `SASL_GSSAPI_PRINCIPAL` cannot be `null` in the way we use the library, right? Also, using `SASL_GSSAPI_PRINCIPAL` together with a `BIND_DN` should trigger a settings validation error.

Reply:
Wouldn't other bind methods not also have their setting keys namespaced similarly to `sasl_gssapi` ? So there would be no ambiguity? To me this is a case of namespaced settings versus conditional settings, as in https://github.com/elastic/elasticsearch/pull/30241 . The conditional setting is harder to validate in the code, compared to the namespaced ones, because the validation of the namespaced ones is just to get the whole namespace and validate them as a group. It's also simpler from a user perspective, because he has fewer settings to worry about.

Reply:
This is just an opinion, do as you think, or ask for other opinions.